### PR TITLE
Add Comba Implementation

### PR DESCRIPTION
**Pull request by AwesomeSeq**
Inspired by the implementation of Gated-DeltaNet, we restructured the Comba operator accordingly. A critical detail is that during the safe_exp computation in Comba, adding a small constant (1e-5) is essential to prevent large deviations during testing. To encapsulate this fix, we introduced the safe_exp_comba function in the utils.op

## Summary by CodeRabbit

- **New Features**
  - Introduced the Comba attention layer and Comba-based language model architecture, now available in the library.
  - Added support for Comba configuration, model, and language modeling head compatible with Hugging Face Transformers.
  - Provided efficient chunked and fused recurrent Comba attention operations, optimized for variable-length sequences and GPU acceleration.

- **Documentation**
  - Updated the README with a news entry announcing the Comba implementation and linking to the related research paper.

- **Tests**
  - Added comprehensive tests to validate the correctness and gradients of Comba attention operations and chunked cumulative sum kernels.

### CODE REVIEW

**Comment by AwesomeSeq** [comment]
Sorry, I made a mistake in the comba.py file. The inputs of the functions chunk_comba and fused_recurrent_comba should include an additional parameter p, for example: 

o, recurrent_state = chunk_comba(
                q=q,
                k=k,
                v=v,
                p=p, （missing term）
                g=g,
                beta=beta,
                initial_state=recurrent_state,
                output_final_state=use_cache,
                cu_seqlens=cu_seqlens,
                use_qk_l2norm_in_kernel=True
            )

**Comment by zhiyuan1i** [comment]
@AwesomeSeq Thank you for your PR!!
Could you please add tests in `tests/test_model.py` and `tests/test_generation.py` ?
And since I have introduece `l2warp` to all models, it would be good if you add some lines in your model, https://github.com/fla-org/flash-linear-attention/commit/82bb6496c1256981749462c7e83db07524e50656

File: fla/__init__.py [hunk 1]
```diff
@@ -61,7 +62,9 @@
     RWKV7ForCausalLM,
     RWKV7Model,
     TransformerForCausalLM,
-    TransformerModel
+    TransformerModel,
+    CombaForCausalLM,
+    CombaModel,
```
Comment (yzhangcs) [inline fla/__init__.py]: @AwesomeSeq Could you sort the model names alphabetically

File: fla/layers/comba.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+        if past_key_values is not None:
+            past_key_values.update(
+                recurrent_state=recurrent_state,
+                conv_state=(conv_state_q, conv_state_k, conv_state_v) if self.use_short_conv else None,
+                layer_idx=self.layer_idx,
+                offset=q_len
+            )
+
+        if self.use_gate:
```
Comment (yzhangcs) [inline fla/layers/comba.py:320]: Let's change this name to `use_output_gate`

File: fla/layers/comba.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+            g = rearrange(self.g_proj(hidden_states), '... (h d) -> ... h d', d=self.head_v_dim)
+            o = self.o_norm(o, g)
+        else:
+            o = self.o_norm(o)
+        o = rearrange(o, 'b t h d -> b t (h d)')
+        o = self.o_proj(o)
+        if attention_mask is not None:
```
Comment (yzhangcs) [inline fla/layers/comba.py:327]: What's the purpose of applying mask upon `o`?

File: fla/models/__init__.py [hunk 4]
```diff
@@ -26,6 +26,7 @@
 from fla.models.rwkv6 import RWKV6Config, RWKV6ForCausalLM, RWKV6Model
 from fla.models.rwkv7 import RWKV7Config, RWKV7ForCausalLM, RWKV7Model
 from fla.models.samba import SambaConfig, SambaForCausalLM, SambaModel
+from fla.models.comba import CombaConfig, CombaForCausalLM, CombaModel
```
Comment (yzhangcs) [inline fla/models/__init__.py]: Would be better if formatting the imports using isort

File: fla/models/__init__.py [hunk 5]
```diff
@@ -52,4 +53,5 @@
     'RodimusConfig', 'RodimusForCausalLM', 'RodimusModel', 'RodimusTokenizer',
     'PaTHAttentionConfig', 'PaTHAttentionForCausalLM', 'PaTHAttentionModel',
     'MesaNetConfig', 'MesaNetForCausalLM', 'MesaNetModel',
```
Comment (yzhangcs) [inline fla/models/__init__.py:56]: also in alphabetic order would be better

File: fla/models/comba/__init__.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# -*- coding: utf-8 -*-
+
+from transformers import AutoConfig, AutoModel, AutoModelForCausalLM
+
+from fla.models.comba.configuration_comba import CombaConfig
+from fla.models.comba.modeling_comba import CombaForCausalLM, CombaModel
+
+AutoConfig.register(CombaConfig.model_type, CombaConfig)
+AutoModel.register(CombaConfig, CombaModel)
+AutoModelForCausalLM.register(CombaConfig, CombaForCausalLM)
```
Comment (yzhangcs) [inline fla/models/comba/__init__.py]: Let's add the `exists_ok` arg by default to avoid some name conflicts.

File: fla/ops/comba/chunk.py [hunk 7]
```diff
@@ -0,0 +1,371 @@
...
+    assert len(beta.shape) == 3, "beta must be of shape [B, T, H] if head_first=False, or [B, H, T] otherwise."
+
+    if p is None:
+        p = k
+    if head_first:
+        raise DeprecationWarning(
+            "head_first is deprecated and will be removed in a future version. "
+            "Please use head_first=False for now instead."
+        )
+        q, k, v, beta, g = map(lambda x: rearrange(x, 'b h t ... -> b t h ...'), (q, k, v, beta, g))
```
Comment (yzhangcs) [inline fla/ops/comba/chunk.py]: This line as well as `[B, H, T, D]` in docstrings can be deleted as `head_first` is deprecated.

File: fla/ops/comba/chunk.py [hunk 8]
```diff
@@ -0,0 +1,371 @@
...
+        p,
+        g,
+        beta,
+        scale,
+        initial_state,
+        output_final_state,
+        cu_seqlens,
+        use_qk_l2norm_in_kernel
+    )
+    if head_first:
```
Comment (yzhangcs) [inline fla/ops/comba/chunk.py]: as above

**Comment by yzhangcs** [comment]
@AwesomeSeq very nice job! I have added some comments, could you resolve them before merging? cc @weigao266

File: README.md [hunk 9]
```diff
@@ -28,6 +28,7 @@ This repo aims at providing a collection of efficient Triton-based implementatio
 * [Acknowledgments](#acknowledgments)
 
 ## News
+- **$\texttt{[2025-06]}$:** 🎉 Add Comba implementation to `fla` ([paper](https://arxiv.org/abs/2506.02475)).
```
Comment (yzhangcs) [inline README.md:31]: Please also add paper links in the table

File: fla/layers/comba.py [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+                activation='silu'
+            )
+        else:
+            warnings.warn(
+                "ShortConvolution is crucial to the performance. "
+                "Do not turn it off, i.e., setting `use_short_conv=False` unless you know what you are doing."
+            )
+        if use_gate:
+            self.g_proj = nn.Linear(hidden_size, self.value_dim, bias=False)
+            self.o_norm = FusedRMSNormGated(self.head_v_dim, eps=norm_eps)
```
Comment (yzhangcs) [inline fla/layers/comba.py]: `sigmoid` is supported in `FusedRMSNormGated`

File: fla/ops/utils/op.py [hunk 11]
```diff
@@ -28,7 +28,10 @@ def div_normal(x, y):
 def safe_exp(x):
     return exp(tl.where(x <= 0, x, float('-inf')))
 
-
+@triton.jit
+def safe_exp_comba(x):
+    return exp(tl.where(x <= 1e-5, x, float('-inf')))
```
Comment (yzhangcs) [inline fla/ops/utils/op.py]: @AwesomeSeq I would suggest not adding this fn in op.py as it is not that general, and we plan to remove all `safe` fns from fla as it is not that safe :-(
